### ISSUE TITLE
MKL causes torch.nn.GELU to run 150x slower on negative inputs

### ISSUE THREAD

**Issue by david-macleod** [pytorch/pytorch#48388]
## 🐛 Bug

The performance of `torch.nn.GELU` on CPU is heavily dependent on the values of the input tensors when MKL is used. When using a tensor containing only (large) negative values the performance is > 150x slower than using a tensor of only postive values. The actual inflection point of the performance cliff appears to be around -14.14, with values smaller than that being slow and tensors larger than that being fast. The negative values appear to be triggering an internal `mkl_vml_kernel_sError` and it seems that the error handler is the process time (possibly due to underflow, hints [here](https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/Profiling-code-that-uses-MKL-mkl-vml-kernel-sError/td-p/765587) and [here](https://code.ihub.org.cn/projects/123/repository/commit_diff?changeset=cd780eb9ec20827a924c658b5960be452797076d)).

## To Reproduce

I run the following code to test GELU speed using input tensors containing either all 100 or all -100
```for i in "100" "-100"; do python -m timeit -s "import torch; x=torch.full((1_000_000,), ${i}.); torch.set_num_threads(1)" -u "msec" "torch.nn.functional.gelu(x)"; done;```

returns 
```
200 loops, best of 5: 0.897 msec per loop
1 loop, best of 5: 145 msec per loop
```
Which shows the negative inputs more than 150x slower than the positive input tensor. However if I hardcode `false` in place of `at::hasMKL()` [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/Activation.cpp#L260) and rebuild pytorch (forcing it to avoid MKL) I get the following result
```
50 loops, best of 5: 5.38 msec per loop
50 loops, best of 5: 5.18 msec per loop
```
So not using MKL is roughly 5x slower than using MKL, but does not have the same performance cliff for (large) negative values

## Expected behavior

Consistency in performance regardless on input value 

## Environment

```
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.18.4

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti
GPU 8: GeForce RTX 2080 Ti
GPU 9: GeForce RTX 2080 Ti

Nvidia driver version: 440.100
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.8.0.4
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.4
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.4
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.4
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.4
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.4
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] pytorch-memlab==0.2.2
[pip3] torch==1.7.0
[pip3] torchaudio==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.8.1
```

## Additional context

- The performance cliff value of -14.14 seems to correspond to `-10 * sqrt(2)`, which has significance in the CdfNorm/GELU calculation
- The py-spy stack trace showing the `mkl_vml_kernel_sError` dominating can be seen here (LHS is `x=100`, RHS is `x=-100`)
![image](https://user-images.githubusercontent.com/17232877/99951620-4ca1e000-2d76-11eb-8c9d-9b392b239b37.png)
- I have also tried calling MKL directly and am able to reproduce the slow/fast results with `vsCdfNorm` (see below). However when I swap in `vsErf` (to emulate the approach to calculating GELU using in PyTorch) the performance is stable. I had assumed `vsCdfNorm` would be calling `vsErf` under the hood but it is a black box.
```c
#include <iostream>
#include <chrono>
#include <mkl.h>

using namespace std;
using namespace std::chrono;

int main(){

    int64_t N = 1000000;
    float X[N];
    float Y[N];
    for (int i = 0; i < N; ++i){
        X[i] = -100;
    }

    auto start = high_resolution_clock::now();
    for (int i = 0; i < 100; ++i){
        vsCdfNorm(N, X, Y);
    }
    auto finish = high_resolution_clock::now();
    duration<double> elapsed = finish - start;
    cout << "Elapsed time: " << elapsed.count() << endl;
    cout << Y[0] << endl;
    return 0;
}
```

cc @VitalyFedyunin @ngimel

**Comment by BIT-silence** [comment]
Hi David, thanks for pointing out this issue. From the Intel MKL documents, it mentioned the [v?CdfNorm](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/vector-mathematical-functions/vm-mathematical-functions/special-functions/v-cdfnorm.html) functions has an extra check for x < underflow, while there is no such check for [v?Erf](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/vector-mathematical-functions/vm-mathematical-functions/special-functions/v-erf.html) functions. I will investigate more and try to resolve this issue with least performance loss.

**Comment by david-macleod** [comment]
Hi @xiaomengy, any updates on this?

**Comment by xiaomengy** [comment]
Sorry for the delay. Based on my investigation, only v?Erf has no VML_STATUS_UNDERFLOW error raised. But v?Erfc and v?CdfNorm both will. I am investigating to see if we can bypass the error by setting some mode or changing to v?Erf with most limited performance influence.

**Comment by xiaomengy** [comment]
It seems https://github.com/pytorch/pytorch/pull/58525 introduced the implementation with mkldnn instead of MKL. Could you try to test the current master to see if there is still the issue?

**Comment by david-macleod** [comment]
Can confirm this recent commit fixes the issue, thanks!
